// @langchain/openai/dist/index.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const require_misc = require('./utils/misc.cjs');
const require_azure = require('./utils/azure.cjs');
const require_base = require('./chat_models/base.cjs');
const require_client = require('./utils/client.cjs');
const require_message_inputs = require('./utils/message_inputs.cjs');
const require_responses = require('./chat_models/responses.cjs');
const require_completions = require('./chat_models/completions.cjs');
const require_index = require('./chat_models/index.cjs');
const require_completions$1 = require('./azure/chat_models/completions.cjs');
const require_responses$1 = require('./azure/chat_models/responses.cjs');
const require_index$1 = require('./azure/chat_models/index.cjs');
const require_llms = require('./llms.cjs');
const require_llms$1 = require('./azure/llms.cjs');
const require_embeddings = require('./embeddings.cjs');
const require_embeddings$1 = require('./azure/embeddings.cjs');
const require_dalle = require('./tools/dalle.cjs');
require('./tools/index.cjs');
const require_custom = require('./tools/custom.cjs');
const require_prompts = require('./utils/prompts.cjs');
const openai = require_rolldown_runtime.__toESM(require("openai"));

exports.AzureChatOpenAI = require_index$1.AzureChatOpenAI;
exports.AzureChatOpenAICompletions = require_completions$1.AzureChatOpenAICompletions;
exports.AzureChatOpenAIResponses = require_responses$1.AzureChatOpenAIResponses;
exports.AzureOpenAI = require_llms$1.AzureOpenAI;
exports.AzureOpenAIEmbeddings = require_embeddings$1.AzureOpenAIEmbeddings;
exports.BaseChatOpenAI = require_base.BaseChatOpenAI;
exports.ChatOpenAI = require_index.ChatOpenAI;
exports.ChatOpenAICompletions = require_completions.ChatOpenAICompletions;
exports.ChatOpenAIResponses = require_responses.ChatOpenAIResponses;
exports.DallEAPIWrapper = require_dalle.DallEAPIWrapper;
exports.OpenAI = require_llms.OpenAI;
Object.defineProperty(exports, 'OpenAIClient', {
  enumerable: true,
  get: function () {
    return openai.OpenAI;
  }
});
exports.OpenAIEmbeddings = require_embeddings.OpenAIEmbeddings;
exports._convertMessagesToOpenAIParams = require_message_inputs._convertMessagesToOpenAIParams;
exports.convertPromptToOpenAI = require_prompts.convertPromptToOpenAI;
exports.customTool = require_custom.customTool;
exports.getEndpoint = require_azure.getEndpoint;
exports.isHeaders = require_azure.isHeaders;
exports.messageToOpenAIRole = require_misc.messageToOpenAIRole;
exports.normalizeHeaders = require_azure.normalizeHeaders;
Object.defineProperty(exports, 'toFile', {
  enumerable: true,
  get: function () {
    return openai.toFile;
  }
});
exports.wrapOpenAIClientError = require_client.wrapOpenAIClientError;
